server/odc-service/src/main/java/com/oceanbase/odc/service/db/schema/DBSchemaIndexService.java

```diff
@@ -0,0 +1,207 @@
+            if (CollectionUtils.isNotEmpty(params.getDatabaseIds())) {
+                List<Database> databases = databaseService.listDatabasesByIds(params.getDatabaseIds());
+                databases.forEach(e -> {
+                    if (Objects.isNull(e.getProject())
+                            || !Objects.equals(e.getProject().getId(), params.getProjectId())) {
+                        throw new NotFoundException(ResourceType.ODC_DATABASE, "id", e.getId());
+                    }
+                });
+                queryDatabaseIds.addAll(databases.stream().map(Database::getId).collect(Collectors.toList()));
+            } else {
+                queryDatabaseIds.addAll(databaseService.listExistDatabaseIdsByProjectId(params.getProjectId()));
+            }
+        } else if (params.getDatasourceId() != null) {
+            Map<Long, Database> id2Database =
+                    databaseService.listExistDatabasesByConnectionId(params.getDatasourceId()).stream()
+                            .collect(Collectors.toMap(Database::getId, e -> e, (e1, e2) -> e1));
+            if (authenticationFacade.currentUser().getOrganizationType() == OrganizationType.INDIVIDUAL) {
+                ConnectionConfig config = connectionService.getBasicWithoutPermissionCheck(params.getDatasourceId());
+                if (!Objects.equals(authenticationFacade.currentUserId(), config.getCreatorId())) {
+                    throw new NotFoundException(ResourceType.ODC_CONNECTION, "id", params.getDatasourceId());
+                }
+                if (CollectionUtils.isNotEmpty(params.getDatabaseIds())) {
+                    queryDatabaseIds.addAll(params.getDatabaseIds().stream().filter(id2Database::containsKey)
+                            .collect(Collectors.toList()));
+                } else {
+                    queryDatabaseIds.addAll(id2Database.keySet());
+                }
+            } else {
+                Set<Long> projectIds = projectService.getMemberProjectIds(authenticationFacade.currentUserId());
+                if (CollectionUtils.isNotEmpty(params.getDatabaseIds())) {
+                    queryDatabaseIds.addAll(params.getDatabaseIds().stream()
+                            .filter(e -> id2Database.containsKey(e) && id2Database.get(e).getProject() != null
+                                    && projectIds.contains(id2Database.get(e).getProject().getId()))
+                            .collect(Collectors.toList()));
+                } else {
+                    queryDatabaseIds.addAll(id2Database.values().stream()
+                            .filter(e -> e.getProject() != null && projectIds.contains(e.getProject().getId()))
+                            .map(Database::getId).collect(Collectors.toList()));
+                }
+            }
+        } else {
+            throw new IllegalArgumentException("projectId or datasourceId is required");
```

Comment:
this branch seems like useless?

Reply:
If the front-end transfer parameters rightly, then this branch is useless. But in back-end codes, we should deal with it.